[DOC] Add example notebook for using aeon distances with sklearn clusterers

SalmanDeveloperz opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at our
contribution guide: https://www.aeon-toolkit.org/en/latest/contributing.html.

Feel free to delete sections of this template if they do not apply to your PR,
avoid submitting a blank template or empty sections.
If you are a new contributor, do not delete this template without a suitable
replacement or reason. If in doubt, ask for help. We're here to help!

Please be aware that we are a team of volunteers so patience is
necessary when waiting for a review or reply. There may not be a quick turnaround for
reviews during slow periods. While we value all contributions big or small, pull
requests which do not follow our guidelines may be closed.
-->

#### Reference Issues/PRs
Fixes #1241
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

#### What does this implement/fix? Explain your changes.
This pull request introduces a new Jupyter Notebook: sklearn_clustering_with_aeon_distances.ipynb. The notebook demonstrates how to integrate aeon's distance metrics with scikit-learn clustering algorithms. It includes:

**Hierarchical Clustering:** Using AgglomerativeClustering with metric="precomputed".
**Density-Based Clustering:** Using DBSCAN and OPTICS with metric="precomputed".
**Spectral Clustering:** Using SpectralClustering with affinity="precomputed" and the inverse of the distance matrix as the similarity matrix.
This addition enhances the clustering documentation, showing how to combine aeon’s distance metrics with widely-used scikit-learn clusterers.

<!--
A clear and concise description of what you have implemented.
-->

#### Does your contribution introduce a new dependency? If yes, which one?
No new dependencies introduced.
<!--
If your contribution does add a dependency, we may suggest adding it as an
optional/soft dependency to keep external dependencies of the core aeon package
to a minimum.
-->

#### Any other comments?

- The notebook has been tested locally, and all cells execute without errors.

- A reference to this notebook has been added to the clustering section of the documentation.

<!--
Any other information that is important to this PR or helpful for reviewers.
-->

### PR checklist

<!--
Please go through the checklist below. Please feel free to remove points if they are
not applicable. To check a box, replace the space inside the square brackets with an
'x' i.e. [x].
-->

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you after the PR has been merged.
- [ ] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [ ] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.


<!--
Thanks for contributing!
-->

SalmanDeveloperz committed:
Add example notebook for using aeon distances with sklearn clusterers

SalmanDeveloperz commented:
Dear maintainers, 
I have created a new notebook and would like to know where to add its reference in the following documentation:

[Clustering Overview](https://www.aeon-toolkit.org/en/latest/examples/clustering/clustering.html#Clustering-notebooks)
[Clustering with sklearn.cluster](https://www.aeon-toolkit.org/en/latest/examples/distances/sklearn_distances.html#Clustering-with-sklearn.cluster)
Could you please guide me on the most appropriate sections in these files to include the reference to the new notebook? I want to ensure it integrates well with the existing content before committing the changes.
Looking forward to your advice.
Best Regards

SebastianSchmidl commented on examples/clustering/sklearn_clustering_with_aeon_distances.ipynb:
```diff
@@ -0,0 +1,274 @@
+{
```
<ul><li>In the beginning, the details on how to use the sklearn estimators are not necessary. Please remove the "with metric=..."-stuff from the introduction and the TOC.</li><li>The TOC does not render correctly</li><li>There is no new content in the _Introduction_-section; can be removed.</li><li> Please rename the section "Loading Data" to "Example Dataset"</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aeon-toolkit/aeon/pull/2511/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

SalmanDeveloperz replied:
Resolved conflicts and updated the notebook as per the review:
- Removed unnecessary "metric=..." details.
- Renamed "Loading Data" to "Example Dataset."
- Deleted the redundant "Introduction" section.
Please let me know if there are any additional changes required.

SebastianSchmidl commented on examples/clustering/sklearn_clustering_with_aeon_distances.ipynb:
```diff
@@ -0,0 +1,274 @@
+{
```
<p>Why so complicated? You could just do</p><p> </p>

<pre class="ql-syntax" spellcheck="false">from aeon.datasets import load_unit_test
X, y = load_unit_test(split="train")
# [...]
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aeon-toolkit/aeon/pull/2511/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

SalmanDeveloperz replied:
Simplified dataset loading as per your feedback. Let me know if further changes are needed!

SebastianSchmidl commented on examples/clustering/sklearn_clustering_with_aeon_distances.ipynb:
```diff
@@ -0,0 +1,274 @@
+{
```
<p>Please link to <a href="https://www.aeon-toolkit.org/en/stable/api_reference/distances.html" target="_blank">https://www.aeon-toolkit.org/en/stable/api_reference/distances.html</a> for an overview of all aeon distances.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aeon-toolkit/aeon/pull/2511/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

SebastianSchmidl replied:
You used the link to the stable-version directly. If you turn it into a relative link, it will point to the correct version.

SalmanDeveloperz replied:
Changed the aeon distances API reference link to a relative link as suggested to ensure it always points to the correct documentation version. Let me know if further adjustments are needed!

SebastianSchmidl commented on examples/clustering/sklearn_clustering_with_aeon_distances.ipynb:
```diff
@@ -0,0 +1,274 @@
+{
```
<p>Please include:</p><ul><li><code>AgglomerativeClustering</code> is, as the name suggests, an agglomerative approach that works by merging clusters bottom up.</li><li>Not all linkage methods can be used with a precomputed distance matrix. <code>single</code>,<code>complete</code>,<code>average</code> , and <code>weighted</code> work with aeon distances.      </li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aeon-toolkit/aeon/pull/2511/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

SalmanDeveloperz replied:
Just updated the "Hierarchical Clustering" section with an explanation of AgglomerativeClustering and the supported linkage methods for precomputed distance matrices, as required. Let me know if further changes are required

SebastianSchmidl replied:
The header "Hierarchical Clustering" exists twice. This is confusing.

SalmanDeveloperz replied:
Thank you for catching this! I sincerely apologize for the confusion caused by the duplicate "Hierarchical Clustering" heading. I have now removed the extra header to improve clarity.
I truly appreciate your patience and feedback.

SebastianSchmidl commented on examples/clustering/sklearn_clustering_with_aeon_distances.ipynb:
```diff
@@ -0,0 +1,274 @@
+{
```
<p><strong>For all estimators</strong>: Please link to the corresponding scikit-learn documentation page.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aeon-toolkit/aeon/pull/2511/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

SalmanDeveloperz replied:
Added links to the corresponding scikit-learn documentation pages for `AgglomerativeClustering`, `DBSCAN`, `OPTICS`, and `SpectralClustering`, as requred. Let me know if there are additional updates required

SebastianSchmidl commented on examples/clustering/sklearn_clustering_with_aeon_distances.ipynb:
```diff
@@ -0,0 +1,274 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #7.    inverse_distance_matrix = 1 /(distance_matrix + epsilon)
</pre><p>Unfortunately, this conversion approach does not preserve the properties of the distance distribution. </p><p>aeon "distances" (they are rather dissimilarities) are in the interval $[0, \infty)$. Thus, we first need to normalize the values to $[0, 1]$, and can, then, simply subtract from 1 to convert to similarities: <code>inverse_distance_matrix = 1 - (distance_matrix / distance_matrix.max())</code> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aeon-toolkit/aeon/pull/2511/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

SalmanDeveloperz replied:
Thanks, for the feedback. Just updated the distance-to-similarity conversion to normalize the distance matrix to [0, 1] and subtract from 1, as required. This ensures that the properties of the distance distribution are preserved. Let me know if further updates are required!

SebastianSchmidl commented on examples/clustering/sklearn_clustering_with_aeon_distances.ipynb:
```diff
@@ -0,0 +1,274 @@
+{
```
<p>You do not cite any of those references. Please add inline references or remove.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aeon-toolkit/aeon/pull/2511/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

SalmanDeveloperz replied:
Removed the references section as it was not cited in the notebook, as suggested.

SebastianSchmidl commented:
> Dear maintainers, I have created a new notebook and would like to know where to add its reference in the following documentation:
> 
> [Clustering Overview](https://www.aeon-toolkit.org/en/latest/examples/clustering/clustering.html#Clustering-notebooks) [Clustering with sklearn.cluster](https://www.aeon-toolkit.org/en/latest/examples/distances/sklearn_distances.html#Clustering-with-sklearn.cluster) Could you please guide me on the most appropriate sections in these files to include the reference to the new notebook? I want to ensure it integrates well with the existing content before committing the changes. Looking forward to your advice. Best Regards

Please refer to the links in [my comment](https://github.com/aeon-toolkit/aeon/issues/1241#issuecomment-2601010189) of the corresponding issue.

SalmanDeveloperz committed:
Resolved conflicts and updated notebook:
- Removed "metric=..." details from TOC and introduction.
- Renamed "Loading Data" to "Example Dataset."
- Deleted redundant Introduction section.

SalmanDeveloperz committed:
Simplified dataset loading using `load_unit_test(split="train")` as suggested.

SalmanDeveloperz committed:
Added this sentence after the introductory line:

"For a comprehensive overview of all available distance metrics in aeon, see the aeon distances API reference."

SalmanDeveloperz committed:
Changes made:-
"AgglomerativeClustering is, as the name suggests, an agglomerative approach that works by merging clusters bottom-up."
Clarified Supported Linkage Methods:

Included the supported linkage methods (single, complete, average, weighted) for precomputed distance matrices.

SalmanDeveloperz committed:
Added links to the scikit-learn documentation pages for all referenced estimators.

SalmanDeveloperz committed:
Updated distance-to-similarity conversion to normalize distances and subtract from 1, ensuring proper preservation of distance distribution.

SalmanDeveloperz committed:
Removed the references section as it was not cited in the notebook, per feedback.

SalmanDeveloperz committed:
Added a reference to the new notebook (sklearn_clustering_with_aeon_distances.ipynb) in the Clustering Overview under Clustering Notebooks.

MatthewMiddlehurst commented:
this is failing the tests currently, please click on details to see why and resolve.

SebastianSchmidl commented:
Could you add a link to your new notebook from the [Clustering-with-sklearn.cluster section](https://www.aeon-toolkit.org/en/latest/examples/distances/sklearn_distances.html#Clustering-with-sklearn.cluster) of the `distances/sklearn_distances`-notebook, too?

SalmanDeveloperz committed:
Updated the aeon distances API reference link to a relative link for version consistency.

SalmanDeveloperz committed:
Removed duplicate "Hierarchical Clustering" header to improve clarity and avoid confusion.

SalmanDeveloperz committed:
Added a reference to the new notebook in the Clustering-with-sklearn.cluster section of sklearn_distances.ipynb.

SalmanDeveloperz commented:
> Could you add a link to your new notebook from the [Clustering-with-sklearn.cluster section](https://www.aeon-toolkit.org/en/latest/examples/distances/sklearn_distances.html#Clustering-with-sklearn.cluster) of the `distances/sklearn_distances`-notebook, too?
>
Inserted the required reference in the Clustering-with-sklearn.cluster section of sklearn_distances.ipynb. Please advise if any further refinements are necessary.

SebastianSchmidl commented:
I think, we also miss a link to this notebook from https://aeon-toolkit--2511.org.readthedocs.build/en/2511/examples.html#clustering

SebastianSchmidl committed:
Merge branch 'main' into add-sklearn-clustering-example

SebastianSchmidl committed:
Automatic `pre-commit` fixes

SalmanDeveloperz committed:
Added a reference to the new sklearn clustering notebook in the Clustering section of examples.md.

SalmanDeveloperz commented:
Added the missing link to the new notebook in the Clustering section of examples.md. Please let me know if any further changes are needed!

SalmanDeveloperz commented:
Hello,
I hope this message finds you well. A soft reminder, I think all the required changes are done and reviewed. Please take a look and let me know if any furture modifications required.
Thanks

SebastianSchmidl commented:
Hi @SalmanDeveloperz,
the CI is still failing. We cannot merge any PRs with failing CI jobs. Please resolve the issues.

SalmanDeveloperz committed:
Fix: Corrected DTW metric in aeon pairwise_distance to resolve CI jobs issue

SalmanDeveloperz committed:
changing matric to method for dtw, trying to resolved the CI Jobs issue.

SalmanDeveloperz committed:
Automatic `pre-commit` fixes

SalmanDeveloperz committed:
Fix AgglomerativeClustering error by replacing 'affinity' with 'metric' to fix CI Job issue.

SalmanDeveloperz committed:
Merge branch 'add-sklearn-clustering-example' of https://github.com/SalmanDeveloperz/aeon into add-sklearn-clustering-example

SalmanDeveloperz committed:
Automatic `pre-commit` fixes

SalmanDeveloperz commented:
Hi @SebastianSchmidl,
I've carefully reviewed the CI failures and made the necessary corrections based on the [aeon distances API reference](https://www.aeon-toolkit.org/en/stable/api_reference/auto_generated/aeon.distances.dtw_distance.html) and a thorough review of the implementation.
After multiple iterations and meticulous debugging, all CI checks have now passed successfully. 
Looking forward to your feedback!

SebastianSchmidl commented on docs/examples.md:
```diff
@@ -165,6 +165,17 @@ Partitional TSCL
 :::
 
+:::{grid-item-card}
+:img-top: examples/clustering/img/sklearn_clustering.png
```
This image is not found. Did you forget to add it to git?

SebastianSchmidl replied:
TY! Not from my side. Let's wait for Matt's response

SebastianSchmidl reviewed (commented):
Nice work 👍

The example plots (matplotlib) are not included in the notebook output, yet. I do not know if this should happen automatically or not.
@MatthewMiddlehurst Are the notebooks built and committed back by our nightly CI builds?

SalmanDeveloperz committed:
Added sklearn clustering example image

SalmanDeveloperz committed:
Added sklearn clustering example image

SalmanDeveloperz committed:
Automatic `pre-commit` fixes

MatthewMiddlehurst commented:
No this is not done automatically, but may be a good idea for those with no output. It would be good to run the notebook so the outputs are displayed here i agree.

https://aeon-toolkit--2511.org.readthedocs.build/en/2511/examples/clustering/sklearn_clustering_with_aeon_distances.html

SalmanDeveloperz commented:
Hi,
The PR is updated as requested. Please review and merge if no further changes are needed. Thanks :)
Best

MatthewMiddlehurst commented:
I see no changes since my comment. You should run the notebook so it has output visible on the website. Make sure you check the readthedocs output under checks.

SalmanDeveloperz committed:
Display notebook outputs and improve clustering accuracy

- Added outputs for Spectral and Hierarchical Clustering plots.
- Fixed oversight per @MatthewMiddlehurst's feedback.
- Refined code for more accurate clustering results.

SalmanDeveloperz committed:
Merge branch 'add-sklearn-clustering-example' of https://github.com/SalmanDeveloperz/aeon into add-sklearn-clustering-example

SalmanDeveloperz committed:
resolve conflict

SalmanDeveloperz committed:
git commit -m "Fix IndentationError in sklearn_clustering_with_aeon_distances.ipynb"

SalmanDeveloperz commented:
Dear @MatthewMiddlehurst,

I sincerely apologize for the oversight in my previous updates. I have now resolved the issues by fixing the merge conflict and correcting the indentation in the notebook. All GitHub Actions checks have passed successfully, and the notebook outputs for both Spectral and Hierarchical Clustering are now visible as requested. I’ve also verified the readthedocs output under checks. Kindly review the changes, and please let me know if any further adjustments are needed. Thank you for your guidance and patience throughout this process :)
Best

MatthewMiddlehurst commented:
None of these plots look informative to me, unfortunately. What is it you want to show with these figures?

Maybe look into how others visualise clustering results (this may be a bit different for time series). There may be some useful functions in `visualisation` you can use.

MatthewMiddlehurst commented on examples/clustering/sklearn_clustering_with_aeon_distances.ipynb:
```diff
@@ -0,0 +1,342 @@
+{
```
<pre class="ql-syntax" spellcheck="false">X = np.vstack(
&nbsp;&nbsp;&nbsp;(
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;np.random.normal(loc=[2, 2], scale=0.5, size=(50, 2)),
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;np.random.normal(loc=[5, 5], scale=0.5, size=(50, 2)),
&nbsp;&nbsp;&nbsp;)
)
</pre><p>Use time series data like you did previously. As mentioned there are unique challenged when it comes to visualising time series, we are not that interested in regular clustering here.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aeon-toolkit/aeon/pull/2511/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

SalmanDeveloperz replied:
Thanks for the feedback. I have updated the example to use time series data instead of the randomly generated clusters. Please check.

SalmanDeveloperz commented:
Hi,
Thank you for your valuable feedback, and I sincerely apologize for the current state of the plots. I’m so sorry for the confusion caused. I’m feeling a bit overwhelmed as my previous attempts to update the visualization have not met the intended goal, and the plots don’t seem informative as you’ve noted.

Before proceeding with further changes, I’d like to seek your guidance. My intention was to demonstrate the separation of time series clusters using Hierarchical Clustering with DTW distance, but the mean-based plots (and my recent shift to time series data) haven’t effectively shown this :(. I noticed the previous 2D data code produced unexpected results locally (in Jupyter and VS Code), though it initially satisfied the output requirement. I’m unsure how to best proceed to align with aeon’s focus on time series clustering.
Please let me know how I can align better with your expectations 🙂. 
Thanks for your patience

MatthewMiddlehurst commented:
Look at the plotting ideas in aeon currently and you may find some ideas. If nothing works, perhaps look at how others plot clustered time series.

SalmanDeveloperz committed:
Re-applied Spectral clustering on time series data using Aeon

MatthewMiddlehurst commented:
Hello, is this still being worked on?

SalmanDeveloperz committed:
Improved DBSCAN clustering visualization to show representative time series per cluster
